feature: add support for SMDDP collectives to smdataparallel runner

vishwakaria opened the pull request:
*Description of changes:*
Add support for SMDDP collectives in PT DDP distribution via a new parameter in the dictionary:
```
communication_options: {
                        "backend": "nccl",  #default value is auto
                       }
```
The distribution will set a configuration parameter called `sagemaker_communication_backend`. If the value is `auto`, we will preload libsmddp which has the Sagemaker optimized implementation of AllReduce. If the value is `nccl`, we will just use the nccl-allReduce implementation.

*Testing done:*
* Added unit tests
* Created a custom docker image and ran successful training

* Related PySDK PR: https://github.com/aws/sagemaker-python-sdk/pull/3490

## Merge Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your pull request._

#### General

- [X] I have read the [CONTRIBUTING](https://github.com/aws/sagemaker-training-toolkit/blob/master/CONTRIBUTING.md) doc
- [X] I used the commit message format described in [CONTRIBUTING](https://github.com/aws/sagemaker-training-toolkit/blob/master/CONTRIBUTING.md#committing-your-change)
- [ ] I have used the regional endpoint when creating S3 and/or STS clients (if appropriate)
- [X] I have updated any necessary documentation, including [READMEs](https://github.com/aws/sagemaker-training-toolkit/blob/master/README.md)

#### Tests

- [X] I have added tests that prove my fix is effective or that my feature works (if appropriate)
- [X] I have checked that my tests are not configured for a specific region or account (if appropriate)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

sandeep-krishnamurthy reviewed (commented):
Thanks.

vishwakaria commented on src/sagemaker_training/smdataparallel.py:
```diff
@@ -432,3 +451,24 @@ def _parse_custom_mpi_options(custom_mpi_options):
+def _validate_smddp_coll_libs_for_preload():
+    """Check if SMDDP collective communication libraries are present in container
+    and have executable permissions"""
+    smddp_lib_paths = [
```
Addressed! Thanks for the review.

satishpasumarthi commented on src/sagemaker_training/smdataparallel.py:
```diff
@@ -241,6 +248,18 @@ def _get_instance_type(self):
+        sm_training_env = json.loads(self._env_vars.get("SM_TRAINING_ENV"))
+        comm_backend = sm_training_env.get("additional_framework_parameters").get(
+            "sagemaker_communication_backend"
+        )
```
can you add a logger statement to dump the `sagemaker_communication_backend` ? This could help us catch if the entry is not set correctly through pysdk.

satishpasumarthi replied:
can you add a logger statement to dump the `sagemaker_communication_backend` ? This could help us catch if the entry is not set correctly through pysdk.

satishpasumarthi commented on src/sagemaker_training/smdataparallel.py:
```diff
@@ -432,3 +451,32 @@ def _parse_custom_mpi_options(custom_mpi_options):
+    """
+    smddp_lib_paths = [
+        "/opt/conda/lib/libsmddp.so",
+        "/opt/conda/lib/libsmddpcoll.so",
```
Could these change in the future? 
Let's make this `smddp_lib_paths`-> `SMDDP_LIB_PATHS` and move to the start of the file.

vishwakaria replied:
These should not change in the near future. 
Will make it a constant.

satishpasumarthi commented on src/sagemaker_training/smdataparallel.py:
```diff
@@ -432,3 +451,32 @@ def _parse_custom_mpi_options(custom_mpi_options):
+            logger.warning("Could not find lib %s", path)
+            logger.warning(
+                "SMDDP collective libraries are not present in the container. "
+                "Please follow <TBA> to set them up on your container."
```
what's `<TBA>` here?

satishpasumarthi commented on test/unit/test_smdataparallel.py:
```diff
@@ -274,7 +293,7 @@ def test_smdataparallel_run_single_node_python(
         async_gather.assert_called_once()
         assert process == async_shell.return_value
-        path_exists.assert_called_with("/usr/sbin/sshd")
+        path_exists.assert_any_call("/usr/sbin/sshd")
```
why was this changed?

vishwakaria replied:
`assert_called_with` checks the latest call to that method. 
Since we added some path exists checks for SMDDP which happen after the sshd checks, this assertion fails. 
Hence, updated it to use `assert_any_call` instead.

satishpasumarthi commented on src/sagemaker_training/smdataparallel.py:
```diff
@@ -241,6 +248,18 @@ def _get_instance_type(self):
+        """Check if communication_backend is set to auto.
+        If yes, enable SMDDP collectives support."""
+        sm_training_env = json.loads(self._env_vars.get("SM_TRAINING_ENV"))
+        comm_backend = sm_training_env.get("additional_framework_parameters").get(
```
This sort of nested get would lead to AttributeError if the first dictionary fails to have the key. Can we break it down ?

satishpasumarthi reviewed (changes requested):
Left few comments

satishpasumarthi commented:
Please follow the commit message style as per https://github.com/aws/sagemaker-training-toolkit/blob/master/CONTRIBUTING.md#committing-your-change 
If you are done with your changes, see if you can squash them into one @vishwakaria

vishwakaria commented:
> Please follow the commit message style as per https://github.com/aws/sagemaker-training-toolkit/blob/master/CONTRIBUTING.md#committing-your-change If you are done with your changes, see if you can squash them into one @vishwakaria

Addressed your comments and squashed all commits. Can you take another look @satishpasumarthi? Thank you.

satishpasumarthi commented on src/sagemaker_training/smdataparallel.py:
```diff
@@ -432,3 +464,35 @@ def _parse_custom_mpi_options(custom_mpi_options):
+    essential for successful training. Thus, if the libraries are absent,
+    we want to avoid training failure by disabling SMDDP Collectives.
+    """
+    smddp_lib_paths = [
```
you can remove this smddp_lib_paths since we defined at top.

vishwakaria replied:
My bad, fixed.

satishpasumarthi reviewed (approved):
LGTM
